### DISCUSSION TITLE
Inprecise description of how stabilityDays works

### DISCUSSION THREAD

**Discussion by MShekow** [renovatebot/renovate#10419] [Request Help] [answered]
Hi! I have the feeling that the docs don't describe the `stabilityDays` configuration option very well. In my mind, the idea (which may be completely wrong) of this config option is to avoid PR/branch-creation-spam for those dependencies where the authors publish many updates (hot fixes) in a very short time interval (e.g. because they are not testing their releases well), and then there is a period of inactivity (for several days or week). Setting `stabilityDays` to e.g. `3` would then only cause the Renovate bot to create a PR/branch if the most recently detected version of a dependency has not changed for 3 days.

However, wouldn't that mean that if some dependency is updated _all the time_ (like Renovate itself, with about 3 releases per day), I'd *never* get updates? I'm pretty sure that this is not the case, it would be a huge problem.

**Comment by HonkingGoose** [comment, marked as answer]
Renovate waits for `stabilityDays` to be reached for **each** separate version before opening a PR.

**Example:**

You have a dependency on `gotta-release-fast`, and you're currently on `0.9.0`, and you set 3 `stabilityDays`, either globally or just for the `gotta-release-fast` package.

`gotta-release-fast` releases a new version each day:

- `1.0.0` day 0
- `1.0.1` day 1
- `1.0.2` day 2

Renovate would then open pull requests to update you, after 3 days have passed for each version, like this:

- `1.0.0` on day 3
- `1.0.1` on day 4
- `1.0.2` on day 5

So Renovate doesn't wait until the `gotta-release-fast` project has not released anything for 3 days. 😉

**Reply by HonkingGoose** [reply]
I've read our docs for `stabilityDays`, and I agree that it's not really clear on the behavior right now... 😄 

I've opened a pull request to improve the `stabilityDays` config option documentation, see #10420.

**Reply by MShekow** [reply]
Thank you for the fast response. For the functionality you are describing, I would have expected the config option to be called `delayPrDays` or similar. One could argue that `stabilityDays`, as implemented right now, is not related to stability - Renovate bot is simply delaying things. What is a concrete use case for this? I don't understand why one would simply want to delay things this way (given that it is unrelated to reducing noise, where you pointed out the solution for that in your other response).

Stability (to me) means: wait until things are stable (= are no longer rapidly changing).

A sensible implementation (to me) would have been something like: "only update to the next *minor* version if there have not been any new *patch*-type version updates (for that new minor version) since `<stabilityDays>` days".

Or one could also make it unspecific to the update types: simply wait until no new versions (of any update type) have been released for `<stabilityDays>` days.

**Reply by HonkingGoose** [reply]
> Thank you for the fast response.

Cheers!

**Use case for `stabilityDays`**

> One could argue that `stabilityDays`, as implemented right now, is not related to stability - Renovate bot is simply delaying things. What is a concrete use case for this?

There's one usecase at least: npm packages can be removed from the npm.js registry within 3 days. If you update straight away with a Renovate bot PR, and the package gets pulled from the registry, you now have a broken package. (Renovate will see this and provide a rollback PR for npm packages). You can prevent the breakage from happening in the first place by waiting 3 `stabilityDays`.

---

**Questions for head honcho @rarkins**

#### `stabilityDays` -> `delayPrDays`

> For the functionality you are describing, I would have expected the config option to be called `delayPrDays` or similar.

I like the idea of calling this functionality `delayPrDays`, I think that does describe a lot better what the current behavior is from the name. What do you think @rarkins? I though the plan was to de-emphasize or drop `stabilityDays` once we have "Merge Confidence" controls that the end-user can use?

> A sensible implementation (to me) would have been something like: "only update to the next _minor_ version if there have not been any new _patch_-type version updates (for that new minor version) since `<stabilityDays>` days".
> 
> Or one could also make it unspecific to the update types: simply wait until no new versions (of any update type) have been released for `<stabilityDays>` days.

That seems like a feature request idea to me, what do you think @rarkins?
I guess they could build this themselves with the scheduling options, so I don't know if you see value in having that feature?

**Reply by MShekow** [reply]
> There's one usecase at least: npm packages can be removed from the npm.js registry within 3 days [...] You can prevent the breakage from happening in the first place by waiting 3 stabilityDays.

Good point. It might even be worthwhile to mention this use case (and potentially others) in the docs. Also, thank you for already creating a PR to update the docs, so that future readers of the docs won't ask the same question I did.

> I guess they could build this themselves with the scheduling options, so I don't know if you see value in having that feature?

If I understand correctly, scheduling is a rather "rigid" schedule that does not consider when a dependency version update was released. I could end up in a situation where I have a schedule, e.g. to update a dependency once per week on Monday early morning, but there is a rush of releases right around that time, e.g. at Monday 1 AM (a buggy release), one at 2 PM (fixing some of the bugs), and a few more on Tuesday and Wednesday (fixing the remaining bugs). Renovate would update the dependency to the most buggy one in this scenario, and I'd be left with that buggy dependency for a whole week.

**Reply by HonkingGoose** [reply]
> Good point. It might even be worthwhile to mention this use case (and potentially others) in the docs. Also, thank you for already creating a PR to update the docs, so that future readers of the docs won't ask the same question I did.

Found the explanation in the docs, it was very well hidden! It's in a completely different section of the docs, no wonder you didn't find it. And I nearly forget we had it at all... 😄 

> Quote from bottom of https://docs.renovatebot.com/merge-confidence/ page
>
> npm packages are never classified as High Confidence until they are at least 3 days old. This is because npm packages less than 72 hours old can be unpublished, which could result in a service impact if you have already updated to it.

I think we need to add this explanation to the `stablityDays` config option docs as well.
**EDIT:** I've added the usecase to the docs now.

---

> If I understand correctly, scheduling is a rather "rigid" schedule that does not consider when a dependency version update was released. I could end up in a situation where I have a schedule, e.g. to update a dependency once per week on Monday early morning, but there is a rush of releases right around that time, e.g. at Monday 1 AM (a buggy release), one at 2 PM (fixing some of the bugs), and a few more on Tuesday and Wednesday (fixing the remaining bugs). Renovate would update the dependency to the most buggy one in this scenario, and I'd be left with that buggy dependency for a whole week.

This seems like a thing that "Merge Confidence" can fix, once that functionality is built. In short it will be a way to tell Renovate to only give you updates for reliable, battle-tested versions, and skip the chaff in between.

Currently Merge Confidence only shows you 4 badges in the update PR, which give you a overview of the state of the version you're upgrading to (Age, Adoption, Passing and Confidence).

The end-goal, I think, is to allow users to set a "confidence level" for package update PRs, either globally or per-package.

Read the [Merge Confidence](https://docs.renovatebot.com/merge-confidence/) docs for information on the current behavior.

**Comment by HonkingGoose** [comment]
To reduce the noise of `gotta-release-fast` you can do something like this in your Renovate config file:

```json
  "packageRules": [
    {
      "matchPackagePatterns": [ "gotta-release-fast" ],
      "groupName": "gotta-release-fast",
      "schedule": ["on the first day of the month"]
    }
  ]
```

In the example above you'd only get a PR once a month, but you can specify any frequency you like.

I recommend you read the [Renovate docs on noise reduction](https://docs.renovatebot.com/noise-reduction/).

I guess you have already read the [Renovate docs on stabilityDays](https://docs.renovatebot.com/configuration-options/#stabilitydays), as you're giving feedback that the description needs to be clearer... 😄

**Reply by HonkingGoose** [reply]
Alternative option, use the "Dependency Dashboard approval workflow" for the `gotta-release-fast` package:

```json
{
  "packageRules": [
    {
      "matchPackagePatterns": ["gotta-release-fast"],
      "dependencyDashboardApproval": true
    }
  ]
}
```

The config above makes it so that the only way the `gotta-release-fast` package will get updated, is by you clicking on the checkbox in the Dependency Dashboard issue.

See [Renovate Docs, dependency dashboard approval option](https://docs.renovatebot.com/configuration-options/#dependencydashboardapproval) for more information.

**Reply by viceice** [reply]
You can also use different stability days per update type.

So for example you can defer minor and major update for 7 days, but defer patch updates only for 3 days.
